From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: nopjmp <kthompson@hey.com>
Date: Fri, 24 Dec 2021 11:14:35 -0600
Subject: [PATCH] Per Tick Flush consolidation

Based on patch by SpottedLeaf

[PATCH] Consolidate flush calls for entity tracker packets

Most server packets seem to be sent from here, so try to avoid
expensive flush calls from them.

This change was motivated due to local testing:

- My server spawn has 130 cows in it (for testing a prev. patch)
- Try to let 200 players join spawn

Without this change, I could only get 20 players on before they
all started timing out due to the load put on the Netty I/O threads.

With this change I could get all 200 on at 0ms ping.

(one of the primary issues is that my CPU is kinda trash, and having
4 extra threads at 100% is just too much for it).

So in general this patch should reduce Netty I/O thread load.

diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 4c6de8a38d7d90ad9cfe0192c2d08d07f349e52a..9bba76909494839d587cdd898830202695bf6985 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -946,7 +946,26 @@ public abstract class MinecraftServer implements ICommandListener, Runnable, IAs
 
                 this.methodProfiler.b();
                 this.methodProfiler.a("tracker");
+                // Dionysus start - Flush consolidation
+                List<net.minecraft.server.PlayerConnection> disabledFlushes = new java.util.ArrayList<>(worldserver.players.size());
+                for (EntityHuman entityHuman : worldserver.players) {
+                    if (entityHuman instanceof EntityPlayer) {
+                        net.minecraft.server.PlayerConnection playerConnection = ((EntityPlayer) entityHuman).playerConnection;
+                        if (playerConnection != null) {
+                            playerConnection.networkManager.disableAutomaticFlush();
+                            disabledFlushes.add(playerConnection);
+                        }
+                    }
+                }
+                try { // Dionysus end - Flush consolidation
                 worldserver.getTracker().updatePlayers();
+                // Dionysus start - Flush consolidation
+                } finally {
+                    for (net.minecraft.server.PlayerConnection playerConnection : disabledFlushes) {
+                        playerConnection.networkManager.enableAutomaticFlush();
+                    }
+                }
+                // Dionysus end - Flush consolidation
                 this.methodProfiler.b();
                 this.methodProfiler.b();
                 worldserver.explosionDensityCache.clear(); // Paper - Optimize explosions
